subsys/bluetooth/audio/has.c

```diff
@@ -433,57 +442,122 @@ static uint8_t preset_found(const struct has_preset *preset, void *user_data)
 	return BT_HAS_PRESET_ITER_STOP;
 }
 
-static int preset_index_compare(const void *p1, const void *p2)
+static void preset_insert(struct has_preset *preset)
 {
-	const struct has_preset *preset_1 = p1;
-	const struct has_preset *preset_2 = p2;
-
-	if (preset_1->index == BT_HAS_PRESET_INDEX_NONE) {
-		return 1;
-	}
+	struct has_preset *tmp, *prev = NULL;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&preset_list, tmp, node) {
+		if (tmp->index > preset->index) {
+			if (prev) {
+				sys_slist_insert(&preset_list, &prev->node, &preset->node);
+			} else {
+				sys_slist_prepend(&preset_list, &preset->node);
+			}
+			return;
+		}
 
-	if (preset_2->index == BT_HAS_PRESET_INDEX_NONE) {
-		return -1;
+		prev = tmp;
 	}
 
-	return preset_1->index - preset_2->index;
+	sys_slist_append(&preset_list, &preset->node);
 }
 
 static struct has_preset *preset_alloc(uint8_t index, enum bt_has_properties properties,
 				       const char *name, const struct bt_has_preset_ops *ops)
 {
-	struct has_preset *preset = NULL;
+	struct has_preset *preset;
+	sys_snode_t *node;
 
-	if (has_preset_num < ARRAY_SIZE(has_preset_list)) {
-		preset = &has_preset_list[has_preset_num];
-		preset->index = index;
-		preset->properties = properties;
+	node = sys_slist_get(&preset_free_list);
+	if (node == NULL) {
+		return NULL;
+	}
+
+	preset = CONTAINER_OF(node, struct has_preset, node);
+	preset->index = index;
+	preset->properties = properties;
 #if defined(CONFIG_BT_HAS_PRESET_NAME_DYNAMIC)
-		utf8_lcpy(preset->name, name, ARRAY_SIZE(preset->name));
+	utf8_lcpy(preset->name, name, ARRAY_SIZE(preset->name));
 #else
-		preset->name = name;
+	preset->name = name;
 #endif /* CONFIG_BT_HAS_PRESET_NAME_DYNAMIC */
-		preset->ops = ops;
-
-		has_preset_num++;
+	preset->ops = ops;
 
-		/* sort the presets in index ascending order */
-		qsort(has_preset_list, has_preset_num, sizeof(*preset), preset_index_compare);
-	}
+	preset_insert(preset);
 
 	return preset;
 }
 
 static void preset_free(struct has_preset *preset)
 {
-	preset->index = BT_HAS_PRESET_INDEX_NONE;
+	bool removed;
+
+	removed = sys_slist_find_and_remove(&preset_list, &preset->node);
+	if (removed) {
+		sys_slist_append(&preset_free_list, &preset->node);
+	}
+}
+
+static struct has_preset *preset_get_prev(const struct has_preset *preset, bool wrap)
+{
+	struct has_preset *tmp, *prev = NULL;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&preset_list, tmp, node) {
+		if (tmp->index == preset->index) {
+			if (prev == NULL && wrap) {
+				sys_snode_t *node = sys_slist_peek_tail(&preset_list);
+
+				prev = CONTAINER_OF(node, struct has_preset, node);
+				if (prev == preset) {
+					return NULL;
+				}
+			}
+			break;
+		}
+		prev = tmp;
+	}
```

Comment:
I have a hard time following this. 

If you want to find the prev, shouldn't it just be 

```c
	SYS_SLIST_FOR_EACH_CONTAINER(&preset_list, tmp, node) {
		if (SYS_SLIST_PEEK_NEXT_CONTAINER(tmp) == preset) {
			return tmp;
		}
	}
```

Why do we need to compare `index` or do wrapping?

Reply:
Alternatively, we could consider using `dlist` if we want to iterate on the list in both directions

Reply:
Thanks for `SYS_SLIST_PEEK_NEXT_CONTAINER`, will use it. For such simple use case it's not worth to use dlist IMO